bin/faucet/src/faucet/mod.rs

```diff
@@ -0,0 +1,422 @@
+                let account = account_file.account;
+                info!(
+                    commitment = %account.commitment(),
+                    nonce = %account.nonce(),
+                    "Faucet not found in the node, using state from file"
+                );
+
+                Ok(account)
+            },
+            Err(err) => Err(err),
+        }
+        .context("fetching faucet state from node")?;
+
+        info!("Fetching genesis header from the node");
+        let genesis_header = rpc_client
+            .get_genesis_header()
+            .await
+            .context("fetching genesis header from the node")?;
+
+        // SAFETY: An empty chain MMR should be valid.
+        let genesis_chain_mmr = ChainMmr::new(
+            PartialMmr::from_peaks(
+                MmrPeaks::new(0, Vec::new()).expect("Empty MmrPeak should be valid"),
+            ),
+            Vec::new(),
+        )
+        .expect("Empty ChainMmr should be valid");
+
+        let data_store = Arc::new(FaucetDataStore::new(
+            account,
+            account_file.account_seed,
+            genesis_header,
+            genesis_chain_mmr,
+        ));
+
+        let public_key = match &account_file.auth_secret_key {
+            AuthSecretKey::RpoFalcon512(secret) => secret.public_key(),
+        };
+
+        let authenticator = Arc::new(BasicAuthenticator::<StdRng>::new(&[(
+            public_key.into(),
+            account_file.auth_secret_key,
+        )]));
+
+        Ok(Self {
+            authenticator,
+            data_store,
+            id,
+            prior_state: VecDeque::new(),
+        })
+    }
+
+    /// Runs the faucet minting process until the request source is closed, or it encounters a fatal
+    /// error.
+    pub async fn run(
+        mut self,
+        mut rpc_client: RpcClient,
+        mut requests: mpsc::Receiver<(MintRequest, oneshot::Sender<(BlockNumber, Note)>)>,
+    ) -> anyhow::Result<()> {
+        let coin_seed: [u64; 4] = random();
+        let rng = RpoRandomCoin::new(coin_seed.map(Felt::new));
+
+        while let Some((request, response_sender)) = requests.recv().await {
+            // Skip doing work if the user no longer cares about the result.
+            if response_sender.is_closed() {
+                tracing::info!(request.account_id=%request.account_id, "request cancelled");
+                continue;
+            }
+
+            match self.handle_request(request, rng, &mut rpc_client).await {
+                // Update local state on success.
+                Ok((delta, block_number, note)) => {
+                    // We ignore the channel closure here as the user may have cancelled the
+                    // request.
+                    let _ = response_sender.send((block_number, note));
+                    // SAFETY: Delta must be valid since it comes from a tx accepted by the node.
+                    self.update_state(&delta).unwrap();
```

Comment:
Would the faucet recover from this failure on restart?

Reply:
On restart the faucet reloads the latest committed state from the node.

This will work in the majority of cases, however this could lead to a mismatch due to transactions still in the mempool. In this case, the faucet would again restart on the first submitted tx (which would fail due to desync). This repeats until all txs have cleared the mempool.

There isn't really a way around that without direct mempool access. And I think this would be a rare case not really worth the bother. At least not yet.